core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CRGenerationPlugin.scala

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (C) 2016-2020 Lightbend Inc. <https://www.lightbend.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cloudflow.sbt
+
+import java.io._
+import java.nio.charset.StandardCharsets._
+
+import spray.json._
+
+import sbt._
+import sbt.Keys._
+
+import scala.util.control.NoStackTrace
+
+import cloudflow.sbt.CloudflowKeys._
+import cloudflow.blueprint.StreamletDescriptor
+import cloudflow.blueprint.StreamletDescriptorFormat._
+import cloudflow.blueprint.deployment.{ CloudflowCR, Metadata, StreamletInstance }
+import cloudflow.blueprint.deployment.CloudflowCRFormat.cloudflowCRFormat
+
+/**
+ * Plugin that generates the CR file for the application
+ */
+object CRGenerationPlugin extends AutoPlugin {
+  final val TEMP_DIRECTORY = new File(System.getProperty("java.io.tmpdir"))
+
+  override def requires =
+    StreamletDescriptorsPlugin && BlueprintVerificationPlugin
+
+  override def projectSettings = Seq(
+    cloudflowApplicationCR := generateCR.dependsOn(verifyBlueprint).value
+  )
+
+  /**
+   * Generate the Application CR from application descriptor generated by verification of
+   * blueprint and the streamlet descriptors generated by the build process. We need the
+   * latter since we need the proper image name associated with every streamlet descriptor
+   * and streamlet deployment.
+   */
+  def generateCR: Def.Initialize[Task[Unit]] = Def.task {
+    // these streamlet descriptors have been generated from the `build` task
+    // if they have not been generated we throw an exception and ask the user
+    // to run the build
+    val log = streams.value.log
+
+    val registry  = cloudflowDockerRegistry.value.get
+    val namespace = cloudflowDockerRepository.value.get
+
+    val streamletDescriptors = imageNamesByProject.value.foldLeft(Vector.empty[StreamletDescriptor]) { (a, e) =>
```

Comment:
Do we assume that we either have all image descriptors or that we have none?
Maybe the user did a `build`, then added a new sub-project and then does `generateCR`. 
Would this hold in that case?

Reply:
@maasg `generateCR` needs a `build` / `buildAndPublish` .. it also needs a `verifyBlueprint` but it can do that as a prerequisite task. But since `build` runs as a sub-project task, it has to be run manually before `generateCR`